ui/modules/environments/environments.js

```diff
@@ -60,65 +65,99 @@ export function ready() {
+    environmentsJsonChanged();
+  };
+
+  dom.tbodyEnvironments.addEventListener('click', (event) => {
+    if (event.target && event.target.tagName === 'TD') {
+      dom.inputEnvironmentName.value = event.target.parentNode.id;
+      updateEnvEditors();
+    }
   });
 
-  document.getElementById('tabEnvironments').onclick = () => {
-    settingsEditor.setValue(JSON.stringify(environments, null, 2), -1);
+  dom.buttonCreateEnvironment.onclick = () => {
+    Utils.assert(dom.inputEnvironmentName.value, 'Please provide an environment name', dom.inputEnvironmentName);
+    Utils.assert(!environments[dom.inputEnvironmentName.value], 'Name already used', dom.inputEnvironmentName);
+    environments[dom.inputEnvironmentName.value] = {
+      api_uri: '',
+    };
+    environmentsJsonChanged();
+  };
+
+  dom.buttonDeleteEnvironment.onclick = () => {
+    Utils.assert(dom.inputEnvironmentName.value, 'No environment selected');
+    Utils.assert(Object.keys(environments).length >= 2, 'At least one environment is required');
+    delete environments[dom.inputEnvironmentName.value];
+    dom.inputEnvironmentName.value = null;
+    environmentsJsonChanged();
   };
 
+  // Ensure that ace editor to refresh when updated while hidden
+  document.querySelector('a[data-bs-target="#tabEnvJson"]').addEventListener('shown.bs.tab', () => {
+    settingsEditor.renderer.updateFull();
+  });
+
+  // Ensure to togle right Ditto user for rest or dev ops api
   document.querySelectorAll('.mainUser,.devOpsUser').forEach((menuTab) => {
     menuTab.addEventListener('click', (event) => {
       API.setAuthHeader(event.target.parentNode.classList.contains('devOpsUser'));
     });
   });
 
-  document.getElementById('environmentSelector').onchange = (event) => {
-    theEnv = event.target.value;
-    activateEnvironment();
-  };
-}
+  document.getElementById('environmentSelector').onchange = notifyAll;
 
-export function togglePinnedThing(evt) {
-  if (evt.target.checked) {
-    current().pinnedThings.push(this.id);
-  } else {
-    const index = current().pinnedThings.indexOf(this.id);
-    if (index > -1) {
-      current().pinnedThings.splice(index, 1);
-    };
-  };
   environmentsJsonChanged();
-};
+}
 
 export function environmentsJsonChanged() {
   localStorage.setItem('ditto-ui-env', JSON.stringify(environments));
 
-  dom.environmentSelector.innerHTML = '';
-  if (theEnv && !current()) {
-    theEnv = null;
+  updateEnvSelector();
+  updateEnvEditors();
+  updateEnvTable();
+  notifyAll();
+}
+
+function updateEnvSelector() {
+  let activeEnvironment = dom.environmentSelector.value;
+  if (!activeEnvironment || !environments[activeEnvironment]) {
+    activeEnvironment = Object.keys(environments)[0];
   };
-  for (const key of Object.keys(environments)) {
+
+  dom.environmentSelector.innerHTML = '';
+  Object.keys(environments).forEach((key) => {
+    // todo: replace with Utils.addDropDownEntries(dom.environmentSelector, [key]);
```

Comment:
What about this "todo"? Shall this be included in the PR?
We usually aim to not leave todos :)

Reply:
No, that is not critical at all. I just removed the todo. Good to know for the future.